### PR TITLE
Cycle orm migrations

### PR DESCRIPTION
**Pull request by roxblnfk**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | no test
| Fixed issues  | #23 

Added new commands:
```bash
  migrate/generate  Generates a migration
  migrate/up        Execute all new migrations
  migrate/down      Rollback last migration
  migrate/list      Print list of all migrations
```

### CODE REVIEW

File: config/common.php [hunk 1]
```diff
@@ -49,6 +52,23 @@
...
     MailerInterface::class => new MailerFactory(),
 
+
+    // Cycle DBAL
+    DatabaseManager::class => new CycleDbalFactory(),
     // Cycle ORM
     ORMInterface::class => new CycleOrmFactory(),
+
+    // Entity dir
+    # todo: replace to class object
```
Comment (samdark) [inline config/common.php]: Do you mean a factory?

Comment (roxblnfk) [reply config/common.php]: I'm not sure. Maybe here needed make a repository. Like EntityFileRepository that will contain paths to all Entities

**Comment by roxblnfk** [comment]
Now i'm satisfied. Please let's review this

File: src/Console/Migration/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+!.gitignore
```
Comment (samdark) [inline src/Console/Migration/.gitignore]: How about moving migrations out of `src/Console`? These could be treated as non-source i.e. be in the `migrations` at the top level or at least could be moved out of `Console`.

**Comment by samdark** [comment]
@roxblnfk I've pushed some minor adjustments. Overall looks good for me except where migrations are located.

File: src/Console/Migration/.gitignore (hunk 2 above)
Comment (roxblnfk) [reply src/Console/Migration/.gitignore]: Maybe you're right. In addition, I also wanted to make a drop cache when applying / rolling back migrations

Comment (roxblnfk) [reply src/Console/Migration/.gitignore]: About the location of the migrations:
Migrations change (add) with Entities. I think they have the right to be in Src along with Entities

Comment (samdark) [reply src/Console/Migration/.gitignore]: OK but, I think, not in Console.

Comment (roxblnfk) [reply src/Console/Migration/.gitignore]: `src/Migration` ?

File: src/Console/Command/MigrateGenerate.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    public function configure(): void
+    {
+        $this->setDescription('Generates a migration');
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $this->cycleHelper->generateMigration($this->migrator, $this->config);
```
Comment (samdark) [inline src/Console/Command/MigrateGenerate.php]: @wolfy-j suggested that [such check is needed](https://github.com/spiral/framework/blob/master/src/Command/Cycle/MigrateCommand.php#L62).

Comment (wolfy-j) [reply src/Console/Command/MigrateGenerate.php]: Cycle will generate migrations based on diff with current DB schema if the schema is not consistent you will generate broken migrations. This is a small nasty issue, so such check is recommended.

Comment (roxblnfk) [reply src/Console/Command/MigrateGenerate.php]: I understand and support this recommendation. Also, it would be nice not to generate empty migrations

Comment (samdark) [reply src/Console/Command/MigrateGenerate.php]: Hmm... Why would empty migrations be generated?

Comment (wolfy-j) [reply src/Console/Command/MigrateGenerate.php]: We are checking generated migrations via Generator observer: https://github.com/spiral/framework/blob/master/src/Command/Cycle/MigrateCommand.php#L78

You are free to copy its code, it also gives you a nice description of what have changed.

Comment (roxblnfk) [reply src/Console/Command/MigrateGenerate.php]: @samdark for example: if Entities not changed then running `migrate/generate` it creates migration contains void methods (`up()` and `down()` )
@wolfy-j ok, I'll see it later, and if necessary, i'll implement the comparison code

Comment (wolfy-j) [reply src/Console/Command/MigrateGenerate.php]: Fixed in 1.0.5

**Comment by wolfy-j** [comment]
I would recommend a way to add empty migration manually. There are always cases where you can't rely on ORM to solve diff.

File: src/Console/Command/MigrateGenerate.php (hunk 3 above)
Comment (samdark) [reply src/Console/Command/MigrateGenerate.php]: One question. How would it work if you write migration manually? Would that still create dummy migration classes for me

Comment (wolfy-j) [reply src/Console/Command/MigrateGenerate.php]: ORM and manual migrations works using the same engine, you can combine them if you want. Basically ORM simply writes migrations for you, the rest is the same.

Comment (wolfy-j) [reply src/Console/Command/MigrateGenerate.php]: https://github.com/cycle/migrations/blob/master/src/GenerateMigrations.php#L87

**Comment by roxblnfk** [comment]
> I would recommend a way to add empty migration manually. There are always cases where you can't rely on ORM to solve diff.

Then you will have to change the code again, because now using \Cycle\Migrations\GenerateMigrations you can not create empty migrations

**Comment by wolfy-j** [comment]
You can not create empty relations for ORM, you still can create migrations directly. Migrator and ORM/Migrations are two different packages.

That's how you can register migration: 
https://github.com/cycle/migrations/blob/master/src/GenerateMigrations.php#L87

And that's how you can generate empty one: 
https://github.com/cycle/migrations/blob/master/src/GenerateMigrations.php#L117

We have a command which allows us to scaffold migrations with some tables and columns pre-defined.

**Comment by roxblnfk** [comment]
@wolfy-j  I just want to avoid code duplication

**Comment by wolfy-j** [comment]
Well, you can always use string body like a classic template.

**Comment by roxblnfk** [comment]
> I'm just saying that empty/user migration is nothing more than just a class in migrations directory.

Yes, that’s obvious. That is why it is useful to check for unapplied migrations before generating new ones. Because if you do not apply the previously generated migration, the new migration will take the all unapplied changes

**Comment by roxblnfk** [comment]
All code migrated to https://github.com/yiisoft/yii-cycle/pull/1

**Pull request closed without merging**